ee/codegen/src/generators/graph-attribute.ts

```diff
@@ -0,0 +1,409 @@
+                values: [
+                  mutableAst,
+                  {
+                    type: "right_shift",
+                    lhs: {
+                      type: "port_reference",
+                      reference: sourcePortContext,
+                    },
+                    rhs: { type: "node_reference", reference: targetNode },
+                  },
+                ],
+              };
+            }
+          } else if (mutableAst.rhs.type === "node_reference") {
+            if (sourceNode && mutableAst.rhs.reference === sourceNode) {
+              return {
+                type: "right_shift",
+                lhs: mutableAst,
+                rhs: { type: "node_reference", reference: targetNode },
+              };
+            }
+          }
+        }
+
+        return;
+      };
+
+      const newMutableAst = addEdgeToGraph(graphMutableAst, null);
+      processedEdges.add(edge);
+
+      if (!newMutableAst) {
+        continue;
+      }
+
+      graphMutableAst = newMutableAst;
+      targetNode.portContextsById.forEach((portContext) => {
+        const edges = this.edgesByPortId.get(portContext.portId);
+        edges?.forEach((edge) => {
+          if (processedEdges.has(edge) || edgesQueue.includes(edge)) {
+            return;
+          }
+          edgesQueue.push(edge);
+        });
+      });
+    }
+
+    return graphMutableAst;
+  }
+
+  /**
+   * Translates our mutable graph AST into a Fern-native Python AST node.
+   */
+  private getGraphAttributeAstNode(mutableAst: GraphMutableAst): AstNode {
+    if (mutableAst.type === "empty") {
+      return python.TypeInstantiation.none();
+    }
+
+    if (mutableAst.type === "node_reference") {
+      const nodeRef = python.reference({
+        name: mutableAst.reference.nodeClassName,
+        modulePath: mutableAst.reference.nodeModulePath,
+      });
+      this.addReference(nodeRef);
+      return nodeRef;
+    }
+
+    if (mutableAst.type === "port_reference") {
+      return python.reference({
+        name: mutableAst.reference.nodeContext.nodeClassName,
+        modulePath: mutableAst.reference.nodeContext.nodeModulePath,
+        attribute: mutableAst.reference.isDefault
+          ? undefined
+          : [PORTS_CLASS_NAME, mutableAst.reference.portName],
+      });
+    }
+
+    if (mutableAst.type === "set") {
+      return python.TypeInstantiation.set(
+        mutableAst.values.map((ast) => this.getGraphAttributeAstNode(ast))
+      );
+    }
+
+    if (mutableAst.type === "right_shift") {
+      const lhs = this.getGraphAttributeAstNode(mutableAst.lhs);
+      const rhs = this.getGraphAttributeAstNode(mutableAst.rhs);
+      if (!lhs || !rhs) {
+        return python.TypeInstantiation.none();
+      }
+      return python.operator({
+        operator: OperatorType.RightShift,
+        lhs,
+        rhs,
+      });
+    }
+
+    return python.TypeInstantiation.none();
+  }
+
+  public write(writer: Writer): void {
+    this.astNode.write(writer);
```

Comment:
This is great! `write` methods should never create new AstNode instances within them. Any AstNode instance that gets created should happen as part of the constructor, which this does 🎉